File: MobileAcceleratorExample/SettingsVC.m
//
//  SettingsVC.m
//  MobileAcceleratorExample
//
//  Created by Brian Salomon on 2/2/16.
//  Copyright © 2016 Akamai. All rights reserved.
//

#import <AkaMap/AkaMap.h>

#import "SettingsVC.h"
#import "AppDelegate.h"
#import "AppSettings.h"

@interface SettingsVC ()

@property (nonatomic, strong) AppSettings *sharedSettings;

@end

@implementation SettingsVC

typedef enum {
	SETTINGS_SERVER_HOST = 0,
    SETTINGS_SDK_API_KEY,
    SETTINGS_SDK_SEGMENTS,
    SETTINGS_WEB_URL,
    NUM_SETTINGS
} settingsRow;

- (void)viewDidLoad {
    [super viewDidLoad];

    self.sharedSettings = [AppSettings sharedInstance];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}


#pragma mark UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return NUM_SETTINGS;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [self.tableSettings dequeueReusableCellWithIdentifier:@"settingsCell"];

    switch (indexPath.row) {
        case SETTINGS_SERVER_HOST: {
            cell.textLabel.text = @"SDK Server (overrides lookup)";
			NSString *serverOverride = appDelegate().akaService.config.serverHostOverride.host;
			cell.detailTextLabel.text = serverOverride ? : @"<use lookup>";
			break;
		}
        case SETTINGS_SDK_API_KEY:
            cell.textLabel.text = @"API Key";
            cell.detailTextLabel.text = self.sharedSettings.serverLicenseKey;
            break;
        case SETTINGS_SDK_SEGMENTS:
            cell.textLabel.text = @"SDK User Segments";
            if (! self.sharedSettings.userSegments) {
                cell.detailTextLabel.text = @"<empty>";
            } else {
                cell.detailTextLabel.text = [self.sharedSettings.userSegments componentsJoinedByString:@","];
            }
            break;
        case SETTINGS_WEB_URL:
            cell.textLabel.text = @"Web View URL";
            cell.detailTextLabel.text = self.sharedSettings.webURL;
            break;
    }

    cell.detailTextLabel.lineBreakMode = NSLineBreakByTruncatingMiddle;
	cell.detailTextLabel.adjustsFontSizeToFitWidth = YES;

    return cell;
}

#pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [self acceptModalInputForField:(settingsRow)indexPath.row];
}

- (void) acceptModalInputForField:(settingsRow)field
{
    NSString *title;
    NSString *message;
    NSString *oldValue;
    switch (field) {
        case SETTINGS_SERVER_HOST:
            title = @"SDK Server Override";
            message = @"(exclude https://)\nClear to use lookup server.";
            oldValue = appDelegate().akaService.config.serverHostOverride.host;
            break;
        case SETTINGS_SDK_API_KEY:
            title = @"SDK API Key";
            oldValue = self.sharedSettings.serverLicenseKey;
            break;
        case SETTINGS_SDK_SEGMENTS:
            title = @"SDK User Segment(s)";
			message = @"Use a comma to separate segments.";
            oldValue = [self.sharedSettings.userSegments componentsJoinedByString:@","];
            break;
        case SETTINGS_WEB_URL:
            title = @"Web View Demo URL";
            oldValue = self.sharedSettings.webURL;
            break;
        default:
            return;
    }

    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];

    [alert addAction:[UIAlertAction actionWithTitle:@"Done" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action){
        UITextField *textField = alert.textFields[0];
        [self handleInput:textField.text forField:field];
        [self.tableSettings reloadData];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [alert addTextFieldWithConfigurationHandler:^(UITextField *textField) {
        textField.placeholder = oldValue;
    }];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void) handleInput:(NSString *)textInput forField:(settingsRow)field
{
    switch (field) {
		case SETTINGS_SERVER_HOST:
			if ([textInput isEqualToString:@""]) {
				appDelegate().akaService.config.serverHostOverride = nil;
			} else {
				NSURL *serverURL = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@", textInput]];
				appDelegate().akaService.config.serverHostOverride = serverURL;
			}
			break;
        case SETTINGS_SDK_API_KEY:
            self.sharedSettings.serverLicenseKey = textInput;
            break;
		case SETTINGS_SDK_SEGMENTS: {
			NSArray *updatedSegments = [textInput componentsSeparatedByString:@","];

			// trim whitespace from ends
			NSMutableArray *trimmedSegments = [NSMutableArray new];
			for (NSString *seg in updatedSegments) {
				NSString *trimmedSeg = [seg stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
				[trimmedSegments addObject:trimmedSeg];
			}
			updatedSegments = [trimmedSegments copy];

			// save to settings for next registration
			self.sharedSettings.userSegments = updatedSegments;

			// update SDK right now
			if (appDelegate().akaService.state != VOCServiceStateNotRegistered) {
				[appDelegate().akaService subscribeSegments:[NSSet setWithArray:updatedSegments]];
			}
			
		} break;
        case SETTINGS_WEB_URL:
            self.sharedSettings.webURL = textInput;
            break;
        default:
            break;
    }
}


@end

File: MobileAcceleratorExampleSwift/SettingsVC.swift
//
//  SettingsVC.swift
//  MobileAcceleratorExampleSwift
//
//  Created by Waazim Reza on 24/05/2016.
//  Copyright © 2016 Akamai. All rights reserved.
//

import Foundation

class SettingsVC: UIViewController, UITableViewDelegate, UITableViewDataSource {

	@IBOutlet weak var tableSettings: UITableView!
	let sharedSettings: AppSettings = AppSettings.sharedInstance

	enum settingsRow: Int {
		case SETTINGS_SERVER_HOST = 0
		case SETTINGS_SDK_API_KEY
		case SETTINGS_SDK_SEGMENTS
		case SETTINGS_WEB_URL
		case NUM_SETTINGS
	}

	// MARK: UITableViewDataSource
	func numberOfSections(in tableView: UITableView) -> Int {
		return 1
	}

	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return settingsRow.NUM_SETTINGS.rawValue
	}

	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell: UITableViewCell = tableSettings.dequeueReusableCell(withIdentifier: "settingsCell")!
		guard let label = cell.textLabel, let detailLabel = cell.detailTextLabel else {
			return cell
		}

		let appDelegate = UIApplication.shared.delegate as! AppDelegate
		let akaConfig = appDelegate.akaService?.config

		if let choice = settingsRow(rawValue: indexPath.row) {
			switch choice {

			case .SETTINGS_SERVER_HOST:
				label.text = "SDK Server (overrides lookup)"
				detailLabel.text = akaConfig?.serverHostOverride?.host ?? "<use lookup>"

			case .SETTINGS_SDK_API_KEY:
				label.text = "API Key"
				detailLabel.text = sharedSettings.serverLicenseKey

			case .SETTINGS_SDK_SEGMENTS:
				label.text = "SDK User Segments"
				detailLabel.text = sharedSettings.userSegments.isEmpty ? "<empty>" : sharedSettings.userSegments.joined(separator: ",")

			case .SETTINGS_WEB_URL:
				label.text = "Web View URL"
				detailLabel.text = sharedSettings.webURL

			default:
				break
			}
		}

		detailLabel.lineBreakMode = .byTruncatingMiddle
		detailLabel.adjustsFontSizeToFitWidth = true

		return cell
	}

	// MARK: UITableViewDelegate
	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		acceptModalInputForField(field: settingsRow(rawValue:indexPath.row)!)
	}

	func acceptModalInputForField(field: settingsRow) {
		var title: String = ""
		var message: String = ""
		var oldValue: String = ""

		let appDelegate = UIApplication.shared.delegate as! AppDelegate
		let akaService = appDelegate.akaService
		let akaConfig = akaService?.config

		switch field {

		case .SETTINGS_SERVER_HOST:
			title = "SDK Server Override"
			message = "(exclude https://)\n Clear to use lookup server."
			if (akaService?.state != VOCServiceState.notRegistered),
				let serverOverrideHost = akaConfig?.serverHostOverride?.host {
				oldValue = serverOverrideHost
			}

		case .SETTINGS_SDK_API_KEY:
			title = "SDK API Key"
			oldValue = sharedSettings.serverLicenseKey

		case .SETTINGS_SDK_SEGMENTS:
			title = "SDK User Segment(s)"
			message = "Use a comma to separate segments."
			oldValue = sharedSettings.userSegments.joined(separator: ",")

		case .SETTINGS_WEB_URL:
			title = "Web View Demo URL"
			oldValue = sharedSettings.webURL

		default:
			break
		}

		let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)

		alert.addAction(UIAlertAction(title: "Done", style: .default, handler: { (alertAction: UIAlertAction!) in
			if let textFields = alert.textFields, textFields.count > 0, let text = textFields[0].text {
				self.handleInput(textInput: text, field: field)
				self.tableSettings.reloadData()
			}
		}))

		alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler:nil))
		alert.addTextField(configurationHandler: { (textField: UITextField!) in
			textField.placeholder = oldValue
		})
		present(alert, animated: true, completion: nil)
	}

	func handleInput(textInput: String, field: settingsRow) {
		let appDelegate = UIApplication.shared.delegate as! AppDelegate
		let akaService = appDelegate.akaService
		let akaConfig = akaService?.config

		switch field {
		case .SETTINGS_SERVER_HOST:
			akaConfig?.serverHostOverride = textInput.isEmpty ? nil :
				URL(string: "https://\(textInput)")

		case .SETTINGS_SDK_API_KEY:
			sharedSettings.serverLicenseKey = textInput

		case .SETTINGS_SDK_SEGMENTS:
			var updatedSegments: [String] = textInput.components(separatedBy:",")

			// trim whitespace from ends
			var trimmedSegments = [String]()
			for seg in updatedSegments {
				let trimmedSeg = seg.trimmingCharacters(in: NSCharacterSet.whitespaces)
				trimmedSegments.append(trimmedSeg)
			}
			updatedSegments = trimmedSegments

			// save to settings for next registration
			sharedSettings.userSegments = updatedSegments

			// update SDK right now
			if (akaService?.state != VOCServiceState.notRegistered) {
				akaService?.subscribeSegments(Set(updatedSegments))
			}

		case .SETTINGS_WEB_URL:
			sharedSettings.webURL = textInput
		default:
			break
		}
	}
}
